### DISCUSSION TITLE
Multiple reference clocks

### DISCUSSION THREAD

**Discussion by cosmicc** [bxparks/AceTimeClock#19] [Ideas]
Is there a way to have multiple reference clocks?

Im using a systemclock with a NTP clock reference and a DS3231 clock backup.
I would like to add a second reference clock (GPS clock I made from the clock class that gets time from my GPS module).

My ultimate goal is to have a GPS reference clock first, if that's not available use the NTP reference clock if available, and keep using the DS3231 as the backup clock reference.

So basically 2 reference clock sources, it uses whichever is available, with priority to the first reference clock.

I was thinking maybe writing a clock class that incorporates both the NTP and GPS reference clock and handles it so it is delivered to acetimeclock as 1 reference source.

Anyone have any ideas?

**Comment by bxparks** [comment]
That an interesting question, I had never considered a configuration with 2 reference clocks. I think the easiest solution is the one you wrote: create a new `CompositeClock` class, which combines your `GpsClock` and `NtpClock` and present it as a single reference clock. The `Clock` interface provides the building blocks to compose these more complicated things.

You will probably encounter a number of implementation hiccups. 

- Failure during initialization: I have found that the the initialization sequence of the NTP clock can be unreliable (the network can fail, the NTP server may be down, etc). But the error condition may not be consistently or easily made available to the end-user. I don't have any experience with GPS clocks, but I imagine that the GPS satellite signal may take extra long time to acquire, and maybe the software times out, but doesn't return the proper error code.
- Failure after initialization: Does `CompositeClock` permanently disable the GPS clock when it detects it as broken, or does it keep retrying at periodic intervals to reconnect to it periodically? Does `CompositeClock` dynamically switch to the other clock if primary `GpsClock` becomes flaky after initialization?

Hope this points you in the right direction.

**Reply by cosmicc** [reply]
It does, thank you. 
I was looking for validation that creating a custom clock class with 2 reference clocks would seem do-able before I started trying to create it.

I figured there would be a bunch of implementation hiccups.  I assumed I would have to write my own handling of the NTP and GPS services being unavailable, keeping track of which is available and periodically trying forever, and dynamically switching between the two so it always delivers one good reference clock back.

So as far as AceTimeClock knows, its just 1 good reference clock and all the NTP and GPS logic is done in the custom clock class.
